File: ProyectoCivitasRuby/lib/juego_texto/controlador.rb
require_relative '../modelo/civitas_juego'
require_relative '../modelo/salidas_carcel'
require_relative '../modelo/operacion_inmobiliaria'

module Civitas
  class Controlador
    
    #Constructor
    def initialize(juego, vista)
      #Atributos de instancia
      @juego = juego
      @vista = vista  
    end
      
    #Métodos públicos    
    def juega
      @vista.set_civitas_juego(@juego)
        
      while(!@juego.final_del_juego) #mientras no acabe el juego, jugará
        @vista.actualizar_vista()
        @vista.pausa()
        operacion = @juego.siguiente_paso()
        @vista.mostrar_siguiente_operacion(operacion)
          
        if(operacion != Operaciones_juego::PASAR_TURNO)
          @vista.mostrar_eventos()
        end
        final = @juego.final_del_juego()
        if(final == false) #si no ha acabado el juego
          if(operacion == Operaciones_juego::COMPRAR)
            respuesta = @vista.comprar
            if(respuesta == Respuestas::SI)
              @juego.comprar()
            end
            @juego.siguiente_paso_completado(operacion)
            
          elsif (operacion == Operaciones_juego::GESTIONAR)
            @vista.gestionar()
            gestion = @vista.get_gestion()
            propiedad = @vista.get_propiedad()
            lista_respuestas = [GestionesInmobiliarias::VENDER, GestionesInmobiliarias::HIPOTECAR, GestionesInmobiliarias::CANCELAR_HIPOTECA, 
              GestionesInmobiliarias::CONSTRUIR_CASA, GestionesInmobiliarias::CONSTRUIR_HOTEL,GestionesInmobiliarias::TERMINAR]
            operinmo = OperacionInmobiliaria.new(lista_respuestas[gestion], propiedad)
              
            case operinmo.gestion
              when GestionesInmobiliarias::VENDER
                @juego.vender(propiedad)

              when GestionesInmobiliarias::HIPOTECAR
                @juego.hipotecar(propiedad)

              when GestionesInmobiliarias::CANCELAR_HIPOTECA
                @juego.cancelar_hipoteca(propiedad)

              when GestionesInmobiliarias::CONSTRUIR_CASA
                @juego.construir_casa(propiedad)

              when GestionesInmobiliarias::CONSTRUIR_HOTEL
                @juego.construir_hotel(propiedad)

              when GestionesInmobiliarias::TERMINAR
                @juego.siguiente_paso_completado(operacion)
            end
              
          elsif(operacion == Operaciones_juego::SALIR_CARCEL)
            salida = @vista.salir_carcel()
            if(salida == Salidas_carcel::TIRANDO)
              @juego.salir_carcel_tirando()
            elsif(salida == Salidas_carcel::PAGANDO)
              @juego.salir_carcel_pagando()
            end
            @juego.siguiente_paso_completado(operacion)
          end
            
        else  #si es el fin del juego
           
          #Mostramos ranking
          top = Array.new();
          top = @juego.ranking
          for i in 0...top.size()
            puts "#{i+1}.  #{top.at(i).nombre}"
          end    
        end    
      end
    end
    
  end
end

File: ProyectoCivitasJava/src/GUI/Controlador.java
/**
 * Proyecto Civitas
 * Sergio Hervás Cobo
 * Javier Suárez Lloréns
 * Programación y Diseño Orientado a Objetos
 * 2º B3 - Grado en Ingeniería Informática
 * E.T.S.I. Informática y de Telecomunicación
 * Universidad de Granada
 */
package GUI;
import civitas.CivitasJuego;
import civitas.Diario;
import civitas.OperacionesJuego;
import civitas.Jugador;
import java.util.ArrayList;
import civitas.OperacionInmobiliaria;
import civitas.GestionesInmobiliarias;
import civitas.SalidasCarcel;

class Controlador {
    private CivitasJuego juego;
    private CivitasView vista;
    
    Controlador(CivitasJuego juego, CivitasView vista){
        this.juego = juego;
        this.vista = vista;
    }
    
    void juega(){
        vista.setCivitasJuego(juego);
        
        while(!juego.finalDelJuego()){
            
            vista.actualizarVista();
            
            //vista.pausa();
            
            OperacionesJuego siguienteop = juego.siguientePaso();
            vista.mostrarSiguienteOperacion(siguienteop);
            
            if (siguienteop != OperacionesJuego.PASAR_TURNO)
                vista.mostrarEventos();
            
            boolean esElFinal = juego.finalDelJuego();
            
            if (!esElFinal){
                
                switch(siguienteop){
                    case COMPRAR:
                        Respuestas resp = vista.comprar();
                        if(resp == Respuestas.SI){
                            juego.comprar();                    
                        }
                        juego.siguientePasoCompletado(siguienteop); 
                        
                break;
                case GESTIONAR:
                    vista.gestionar();
                    int gestion = vista.getGestion();
                    int propiedad = vista.getPropiedad();
                    
                    OperacionInmobiliaria operInmo = new OperacionInmobiliaria(GestionesInmobiliarias.values()[gestion], propiedad);

                        if (operInmo.getGestiones() == GestionesInmobiliarias.CANCELAR_HIPOTECA){
                            juego.cancelarHipoteca(propiedad);
                        }
                        else if (operInmo.getGestiones() == GestionesInmobiliarias.VENDER){
                            juego.vender(propiedad);
                        }
                        else if (operInmo.getGestiones() == GestionesInmobiliarias.HIPOTECAR){
                            juego.hipotecar(propiedad);
                        }
                        else if (operInmo.getGestiones() == GestionesInmobiliarias.CONSTRUIR_CASA){
                            juego.construirCasa(propiedad);
                        }
                        else if (operInmo.getGestiones() == GestionesInmobiliarias.CONSTRUIR_HOTEL){
                            juego.construirHotel(propiedad);
                        }
                        else if(operInmo.getGestiones() == GestionesInmobiliarias.TERMINAR){
                            juego.siguientePasoCompletado(siguienteop);
                        }     
                break;          
                case SALIR_CARCEL:
                            SalidasCarcel salida = vista.salirCarcel();
                            if (salida == SalidasCarcel.PAGANDO){
                                juego.salirCarcelPagando();
                            }
                            else if(salida == SalidasCarcel.TIRANDO){
                                juego.salirCarcelTirando();
                            }
                            juego.siguientePasoCompletado(siguienteop);
                break;
                            
                }
            }

            else{
                //Ranking ahora tiene la visibilidad pública
                ArrayList<Jugador> top = juego.ranking();
                int puesto = 0;
                Diario.getInstance().ocurreEvento("\n\nUn jugador ha caído en la bancarrota.\n\n");
                Diario.getInstance().ocurreEvento("\n#####--RANKING--#####");
                for ( int k = 0; k < top.size(); k++){
                    puesto++;
                    Diario.getInstance().ocurreEvento("TOP: " + puesto + ":  " + top.get(k).toString());
                }
                vista.actualizarVista();
            }
        }
    }      
  }
